### PR TITLE
Ed/move loss

### LINKED ISSUE #30: don't recompute variance from scratch
We're doing almost twice as much computation as necessary here because `qz_m` and `qz_v` have most of their layers in common. 
https://github.com/YosefLab/scVI-dev/blob/f4e7d90275a447f83219a3bc22366cd94c7341ee/scvi/scvi.py#L118-L119

Here too:
https://github.com/YosefLab/scVI-dev/blob/f4e7d90275a447f83219a3bc22366cd94c7341ee/scvi/scvi.py#L122-L123

### LINKED ISSUE #31: set encoder/decoder parameters to match those of VAE
Probably we meant to set the encoder's and decoder's settings (e.g. `n_hidden`) to match the parent. e.g. `n_hidden=n_hidden`, not `n_hidden = 128`.

https://github.com/YosefLab/scVI-dev/blob/f4e7d90275a447f83219a3bc22366cd94c7341ee/scvi/scvi.py#L36-L37

### PR DESCRIPTION
**Pull request by Edouard360**
Resolved issues:

* Moving loss to VAE
* Don't recompute variance from scratch: resolves #30
* Set encoder/decoder parameters to match those of VAE: fixes #31
* Use register buffer for px_r (when dispersion mode is “gene”)
* Only call cuda twice (in dataset and in run_benchmarks after vat init)
* Remove requires_grad in Variable (default option)

Unresolved issues:

* dtype is not completely removed because of kl_ponderation.
* Also register_buffer creates a tensor, but it is necessary to call Variable(self.px_r) afterwards for the operations (there might be a cleaner solution ?)

### CODE REVIEW

File: scvi/dataset/dataset.py [hunk 1]
```diff
@@ -37,23 +32,24 @@ def __init__(self, Xs):
                            torch.from_numpy(i * np.ones((X.size()[0], 1))).type(torch.FloatTensor)),
                           dim=1)]
 
-        self.X = torch.cat(new_Xs, dim=0)  # .type(dtype)
+        self.X = torch.cat(new_Xs, dim=0)
+        if torch.cuda.is_available():  # TODO: check valid
+            self.X.cuda()
```
Comment (jeff-regier) [inline scvi/dataset/dataset.py]: In `dataset.py`, as in `scvi.py` we probably don't want any calls to `.cuda()` or `cuda.is_available`. On this line, for instance, the whole dataset gets transferred to the GPU. But GPUs don't have a lot of memory. Better to transfer the data from the CPU to the GPU in `train.py`, right where you call `Variable` (with both for loops).
